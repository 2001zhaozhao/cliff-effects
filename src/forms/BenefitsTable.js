// REACT COMPONENTS
import React from 'react';
import {
  Table,
  Header
} from 'semantic-ui-react';

// CUSTOM COMPONENTS
// Both the table and graph should just be added to a results page, but
// this will do for now
import ResultsGraph from '../resultsGraph';

// BENEFIT LOGIC
import { getSNAPBenefits } from '../programs/state/massachusetts/snap';
import { getHousingBenefit } from '../programs/state/massachusetts/housing';


const getSignSymbol = function ( num ) {
  if ( num > 0 ) { return '+'; }
  else if ( num < 0 ) { return '-'; }
  else { return ''; }
};  // End getSignSymbol()


const BenefitsTable = function ( props ) {

  var client        = props.client,
      currentClient = { ...client };
  currentClient.futureEarnedIncomeMonthly = currentClient.currentEarnedIncomeMonthly;

  var SNAPBenefitCurrent  = Math.round( getSNAPBenefits( currentClient ).benefitValue * 12 ),
      SNAPBenefitFuture   = Math.round( getSNAPBenefits( client ).benefitValue * 12 ),
      SNAPDiff            = SNAPBenefitFuture - SNAPBenefitCurrent,
      sec8BenefitCurrent  = Math.round( getHousingBenefit( currentClient ).benefitValue * 12 ),
      sec8BenefitFuture   = Math.round( getHousingBenefit( client ).benefitValue * 12 ),
      sec8Diff            = sec8BenefitFuture - sec8BenefitCurrent,
      totalBenefitCurrent = SNAPBenefitCurrent + sec8BenefitCurrent,
      totalBenefitFuture  = SNAPBenefitFuture + sec8BenefitFuture,
      totalDiff           = SNAPDiff + sec8Diff,
      incomeCurrent       = Math.round( client.currentEarnedIncomeMonthly * 12 ),
      incomeFuture        = Math.round( client.futureEarnedIncomeMonthly * 12 ),
      incomeDiff          = incomeFuture - incomeCurrent,
      netCurrent          = totalBenefitCurrent + incomeCurrent,
      netFuture           = totalBenefitFuture + incomeFuture,
      netDiff             = totalDiff + incomeDiff;

console.log(SNAPDiff);



const cellStyle = function(entry) {

  const   basicCellStyle    = {textAlign: 'right'}
        , columnHeaderStyle = {background: 'rgba(0, 128, 128, 1)',
                              color: 'white',
                              fontSize: '1.3em',
                              fontWeight: 900,
                              textAlign: 'center',
                              borderRadius: 'inherit',
                              letterSpacing: '0.02em'
                              }
        , rowHeaderStyle    = {fontSize: '1.2em',
                              fontWeight: 700,
                              textAlign: 'left'}
        , totalsRowStyle    = {borderTop: "2px solid rgba(0, 128, 128, 1)"};


    switch(entry) {
        case 'basicCellStyle': 
          return basicCellStyle; 
        case 'columnHeaderStyle':
          return columnHeaderStyle; 
        case 'rowHeaderStyle' :
          return  Object.assign(basicCellStyle, rowHeaderStyle);
        case 'totalsRowStyle':
          return {function(){
                    Object.assign(basicCellStyle, totalsRowStyle);
                    }
                  };
        default: 
          return basicCellStyle;
                  }
  };


const Cell = function(props) {

  return (
    <Table.Cell width={3} style={cellStyle(props.style)}> {props.value} </Table.Cell>
  );
};



const Row = function(props) {
  const numbers = props.numbers;
  const cell = numbers.map((D) =>
    <Cell key={D.index}
        value={D} />
  );

  return (
    <Table.Row>
      {cell}
    </Table.Row>
  );
}

const numbers = ['Benefit', 'Current Benefits', 'New Estimate', 'Difference'];
const numbers1 = ['SNAP', SNAPBenefitCurrent, SNAPBenefitFuture, Math.abs(SNAPDiff)];
const numbers2 = ['Section 8 Housing', sec8BenefitCurrent, sec8BenefitFuture, Math.abs(sec8Diff)];
const numbers3 = ['Total Benefits', totalBenefitCurrent, totalBenefitFuture, Math.abs(totalDiff)];
const numbers4 = ['Income', incomeCurrent, incomeFuture, Math.abs(incomeDiff)];
const numbers5 = ['Net Total', netCurrent, netFuture, Math.abs(netDiff)];




  return (
    <wrapper>


      <Table celled borderRadius='0.25em 0.25em 0 0'>
        <Table.Body >
            <Row numbers={numbers} />
            <Row numbers={numbers1} />
            <Row numbers={numbers2} />
            <Row numbers={numbers3} />
            <Row numbers={numbers4} />
            <Row numbers={numbers5} />
        </Table.Body>
      </Table>

{/*
      <Table celled>
      <Table.Body>
       <Table.Header>  Removed for formatting purposes temporarily}
          <Table.Row textAlign='center' style={columnHeaderStyle}>
            <Table.Cell width={3}>Benefit</Table.Cell>
            <Table.Cell width={3}>Current Benefits</Table.Cell>
            <Table.Cell width={3}>New Estimate</Table.Cell>
            <Table.Cell width={3}>Difference</Table.Cell>
        </Table.Row>
        </Table.Header>
      </Table.Body>
        <Table.Body>
            <Table.Row>
              <Table.Cell textAlign='left'><Header as='h4'>SNAP</Header></Table.Cell>
              <Table.Cell textAlign='right'>${SNAPBenefitCurrent} / year</Table.Cell>
              <Table.Cell textAlign='right'>${SNAPBenefitFuture} / year</Table.Cell>
              <Table.Cell textAlign='right'>{ getSignSymbol(SNAPDiff) } ${Math.abs(SNAPDiff)} / year</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell textAlign='left'><Header as='h4'>Section 8 Housing</Header></Table.Cell>
              <Table.Cell textAlign='right'>${sec8BenefitCurrent} / year</Table.Cell>
              <Table.Cell textAlign='right'>${sec8BenefitFuture} / year</Table.Cell>
              <Table.Cell textAlign='right'>{ getSignSymbol(sec8Diff) } ${Math.abs(sec8Diff)} / year</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='left'><Header as='h4'>Total Benefits</Header></Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>${totalBenefitCurrent} / year</Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>${totalBenefitFuture} / year</Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>{ getSignSymbol(totalDiff) } ${Math.abs(totalDiff)} / year</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell textAlign='left'><Header as='h4'>Income</Header></Table.Cell>
              <Table.Cell textAlign='right'>${incomeCurrent} / year</Table.Cell>
              <Table.Cell textAlign='right'>${incomeFuture} / year</Table.Cell>
              <Table.Cell textAlign='right'>{ getSignSymbol(incomeDiff) } ${Math.abs(incomeDiff)} / year</Table.Cell>
            </Table.Row>
            <Table.Row style={{border: 'none'}}>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='left'><Header as='h4'>Net Total</Header></Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>${netCurrent} / year</Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>${netFuture} / year</Table.Cell>
              <Table.Cell width={3} style={rowHeaderStyle} textAlign='right'>{ getSignSymbol(netDiff) } ${Math.abs(netDiff)} / year</Table.Cell>
            </Table.Row>
        </Table.Body>

      </Table>
    )*/}

      <br/>
      <br/>

      <ResultsGraph {...props}/>
    </wrapper>
  );

};  // End BenefitsTable(<>)


export {
	BenefitsTable
};
