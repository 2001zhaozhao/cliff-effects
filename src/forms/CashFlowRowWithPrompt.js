import React from 'react';
import { Form } from 'semantic-ui-react';

import { CashFlowRow } from './formHelpers';

const NumberField = (props) => (
  <CashFlowRow {...props} type={'income'} />
);

/**
 * Prompt the user before presenting a CashFlowRow
 * 
 * @param {object} props
 * @param {string} props.generic - The key of the value being set.
 * @param {string} props.promptLabel - Label for initial prompt.
 * @param {string} props.children - Label for fields updating the value.
 * 
 * @extends React.Component
 * @see CashFlowRow
 */
class CashFlowRowWithPrompt extends React.Component {
  constructor(props) {
    super(props);
    
    const value = props.timeState[props.generic];
    this.state = {
      showField: value !== 0,
      storedValue: value
    }
  }

  handleChange = (evt, inputProps) => {
    if(inputProps.value === 'Yes') {
      this.showField(evt);
    } else {
      this.hideField(evt);
    }
  }

  hideField = evt => {
    const { generic, setClientProperty, timeState } = this.props;

    this.setState({
      showField: false,
      storedValue: timeState[generic]
    });

    setClientProperty(evt, {
      name: generic,
      value: 0
    });
  }

  showField = evt => {
    const { generic, setClientProperty } = this.props;
    const { storedValue } = this.state;

    setClientProperty(evt, {
      name: generic,
      value: storedValue
    });

    this.setState({
      showField: true
    });
  }

  render() {
    const { promptLabel, ...rest } = this.props;
    const { showField } = this.state;

    return (
      <div>
        <Form.Group inline>
          <label>{promptLabel}</label>
          <Form.Radio label={'Yes'} value={'Yes'} checked={showField} onChange={this.handleChange} />
          <Form.Radio label={'No'} value={'No'} checked={!showField} onChange={this.handleChange} />
        </Form.Group>
        
        {showField && <NumberField {...rest} />}
      </div>
    );
  }
}

export default CashFlowRowWithPrompt;
