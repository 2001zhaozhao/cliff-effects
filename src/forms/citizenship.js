// REACT COMPONENTS
import React from 'react';
import { Form, Radio } from 'semantic-ui-react';

// PROJECT COMPONENTS
import {
  FormPartsContainer
} from './formHelpers';


/** @todo description
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const CitizenshipContent = ({ setClientProperty, client }) => {

  return (
    <wrapper className={'field-aligner'}>
      <Form.Field>
        <Radio
          label='US Citizen / National'
          name='citizenshipStatus'
          value='citizen'
          checked={client.citizenshipStatus === 'citizen'}
          onChange={setClientProperty}
        />
      </Form.Field>
      <br/>
      <Form.Field>
        <Radio
          label='Lawfully present immigrant / AWSS'
          name='citizenshipStatus'
          value='immigrant'
          checked={client.citizenshipStatus === 'immigrant'}
          onChange={setClientProperty}
        />
      </Form.Field>
      <br/>
      <Form.Field>
        <Radio
          label="Don't Know"
          name='citizenshipStatus'
          value='unknown'
          checked={client.citizenshipStatus === 'unknown'}
          onChange={setClientProperty}
        />
      </Form.Field>
    </wrapper>
  );  // end return

};  // End CitizenshipContent()

/** @todo description
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
// `props` is a cloned version of the original props. References broken.
const CitizenshipStep = function ( props ) {

  return (
    <Form className='citizenship-form'>
      <FormPartsContainer
        title     = {'Citizenship Status'}
        clarifier = {'Select your citizenship status.'}
        left      = {{name: 'Previous', func: props.previousStep}}
        right     = {{name: 'Next', func: props.nextStep}}>
          <CitizenshipContent setClientProperty={props.setClientProperty} client={props.client} />
      </FormPartsContainer>
    </Form>
  );

};  // End CitizenshipStep()

export { CitizenshipStep };
