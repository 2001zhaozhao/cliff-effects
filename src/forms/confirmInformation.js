// REACT COMPONENTS
import React from 'react';
import { Header } from 'semantic-ui-react';

// PROJECT COMPONENTS
import { FormPartsContainer } from './formHelpers';

const ConfirmInformation = (props) => (
		<FormPartsContainer
			title     = 'Confirm Information'
			clarifier = ''
			left      = {{name: 'Previous', func: props.previousStep}}
			right     = {{name: 'Next', func: props.nextStep}}>
			  <div>
				<Header as='span' color='teal'>Current Benefits</Header> 
				{props.client.hasSnap ? "SNAP, ": ""}
				{props.client.hasHousing ? "Section 8, ": ""}
				{props.client.hasMassHealth ? "MassHealth": ""}<br/>
				<Header as='span' color='teal'>Household Size</Header> {props.client.householdSize}<br/>
				<Header as='span' color='teal'>Household Annual Income</Header> {props.client.annualIncome}<br/>
				<Header as='span' color='teal'>Citizenship Status</Header> {props.client.citizenshipStatus == "citizen" ? "US Citizen" : "Resident" }<br/>
				<Header as='span' color='teal'>MassHealth</Header> {props.client.hasMassHealth ? "Yes, has qualifying conditions" : "No qualifying conditions"  }
			  </div>
      </FormPartsContainer>
	  )

export { ConfirmInformation };