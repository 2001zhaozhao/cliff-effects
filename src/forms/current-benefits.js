// REACT COMPONENTS
import React from 'react';
import { Form } from 'semantic-ui-react';

// PROJECT COMPONENTS
import {
  FormPartsContainer, MassiveToggle
} from './formHelpers';


/** @todo description
* 
* @todo Add "vertical list of options" creator that will create a list of fields using the `.field-aligner` class
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const CurrentBenefitsContent = (props) => {

  var client = props.pageState;

  return (
    <wrapper className={'field-aligner'}>
      <MassiveToggle
        label={{ children: client.hasSnap ? <strong>SNAP</strong> : 'SNAP' }}
        checked={client.hasSnap}
        onChange={props.storeChecked}
        name='hasSnap' />
      <br/>
      <MassiveToggle
        label={{ children: client.hasHousing ? <strong>Section 8 Housing</strong> : 'Section 8 Housing' }}
        checked={client.hasHousing}
        onChange={props.storeChecked}
        name='hasHousing' />
      <br/>
      <MassiveToggle
        label={{ children: client.hasMassHealth ? <strong>MassHealth</strong> : 'MassHealth' }}
        checked={client.hasMassHealth}
        onChange={props.storeChecked}
        name='hasMassHealth' />
    </wrapper>
  );  // end return

};  // End CurrentBenefitsContent()

/** @todo description
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
// `props` is a cloned version of the original props. References broken.
const CurrentBenefitsStep = (props) => {

  return (
    <Form size='massive' className='household-size-form'>
      <FormPartsContainer
        title     = {'Current Benefits'}
        clarifier = {'Select the benefits you currently receive.'}
        right     = {{name: 'Next', func: props.nextStep}}>
          <CurrentBenefitsContent {...props} />
      </FormPartsContainer>
    </Form>
  );

};  // End CurrentBenefitsStep()

export { CurrentBenefitsStep };
