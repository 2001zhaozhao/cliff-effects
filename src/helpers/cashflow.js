/** For all income and general cash flow getters and calculators
* 
* @todo Add checks for properties with messages for mistakes
* @todo Deal with expiration dates
* @todo Prune values now understood to be non-standard
*/

// DATA
import { UNEARNED_INCOME_SOURCES } from '../data/state/massachusetts/name-cores';

// ==================================
// DEDUCTIONS
// ==================================
/**
* Calculates various deductions based on form input
* Everything is monthly
* or
* var getDeduction = {};
* getDeduction.medical = function ( client, timeframe ) {};
* Example of use: getDeduction[ 'medical' ]( client, timeframe );
*/
var deductions = {};

/** 
* @todo add table to data: (@see ยง364.400(A) -
* look up table for standard deductions)
*/
deductions.standard = function ( client, timeframe ) {
};  // End deductions.standard()


/** 
* @todo add table to data: (@see ยง364.400(B) -
* percent amount times earned income)
*/
deductions.earnedIncome = function ( client, timeframe ) {
};  // End deductions.earnedIncome()


/** 
* @todo Find source of these amounts add add to data.
* Standard $155 deduction allowed if medical expenses are $35-$190/month.
* Actual amount minus $35' allowed if over $190/mo.
*/
deductions.medical = function ( client, timeframe ) {
};  // End deductions.medical()


/** Dependent = child or disabled adult
* Sum of:
* Direct Care Costs
* Before- and After-School Programs
* Transportation Costs
* Other
*/
deductions.dependentCare = function ( client, timeframe ) {
  return client[ timeframe + 'ChildSupportPaidOutMonthly' ];
};  // End deductions.dependentCare()


/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
deductions.handicappedAssistance = function ( client, timeframe ) {
};  // End deductions.handicappedAssistance()


/** Just child support. Here for consistency and validation */
deductions.childSupportPayments = function ( client, timeframe ) {
	// For now, no checks or anything. Fragile, but gets the job done.
	return client.futureChildSupportOutMonthly || client.currentChildSupportOutMonthly || 0;
};  // End deductions.childSupportPayments()


// ==================================
// STRAIGHT UP INCOME
// ==================================

/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const getGrossUnearnedIncomeMonthly = function ( client, timeframe ) {
  return sumCashflow( client, timeframe, UNEARNED_INCOME_SOURCES );
};  // End getGrossUnearnedIncomeMonthly()


/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const sumCashflow = function ( client, timeframe, props ) {
  var sum = 0;
  for (let namei = 0; namei < props.length; namei++) {
    sum += toCashflow( client, timeframe, props[ namei ] );  // if that prop exists, add it
  };
  return sum;
};  // End sumCashflow()


/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const toCashflow = function ( client, timeframe, prop ) {
  return client[ timeframe + prop + 'Monthly' ] || 0;
};  // End toCashflow()


/**
* NON-FUNCTIONAL - different programs calculate gross income
* differently.
* 
* Previous Description:
* Calculates the gross monthly income based on client values
* This is the household gross monthly income.
* Derived from excel formulas from download made available here:
* {@link https://www.masslegalservices.org/content/online-snap-calculator}.
* From 2006 ~~Another source, implying something else:
* {@link https://www.hudexchange.info/resources/documents/incomeresidentrentcalc.pdf}~~
* 
* @todo Add checks for properties with messages for mistakes
* @todo Does this return an instance of `Result`?
* 
* @function
* @todo @param
* `timeframe` can be either 'current' or 'future'
* 
* @todo @returns
*/
const getSimpleGrossIncomeMonthly = function ( client, timeframe ) {
  /** @todo Add checks, validtators, and error messages */
  /** @todo Include info like "only count up to $480/12 of what full time students make" */
  var earned    = client[ timeframe + 'EarnedIncomeMonthly' ],
      unearned  = getGrossUnearnedIncomeMonthly( client, timeframe ),
      comingIn  = earned + unearned,

      /** @todd Store this at some point */
      // client[ timeframe + 'GrossIncomeMonthly' ] = total;

      /** @todo This document {@link https://www.masslegalservices.org/content/online-snap-calculator}
      * implies that there are deductions to be removed from income in order to calculate
      * gross income. Is that a mistake? Is that just for SNAP? */
      // total     = comingIn - deductions.childSupportPayments( client, timeframe );

      total     = comingIn;
	return total;

};  // End getSimpleGrossIncomeMonthly()


// from 2006 ~~Source: https://www.hudexchange.info/resources/documents/incomeresidentrentcalc.pdf~~
const getEarnedIncomeDisregardMonthly = function ( client, timeframe ) {
  // Annual
  // Ask Kristin if this is something she wants to include in the prototype
  // Ask Kristin if disability stuff is desired in the prototype
  return 0;
};  // End getEarnedIncomeDisregardMonthly()


// from 2006 ~~Source: https://www.hudexchange.info/resources/documents/incomeresidentrentcalc.pdf~~
// NON FUNCTIONAL - This is actually just for Section 8
const getAllowancesMonthly = function ( client, timeframe ) {
  // Calculations need annual amounts

  // or use the function?
  var yearlyGross = client[ timeframe + 'GrossIncomeMonthly' ] * 12;


  /** @todo Find source of data for monetary amounts and store as data */
  var sum = 0;  // Allowances
  /** 24CFR5.611(a) */
  sum += 480 * client[ timeframe + 'NumOfDependents' ] || 0;
  // head, spouse, or sole member is >=62 or is handicapped/disabled. ONLY ONE DEDUCTION PER FAMILY/HOUSEHOLD PER YEAR
  if ( client[ timeframe + 'HasDisabledOrElderly' ] ) { sum += 400; }

  // ---- Child care ---- \\
  /** @todo Already gathered from depedant care? That's for disabled or elderly members too, though...
  * @todo Validate in input that it's not more than earned income.
  * @todo Can't be expenses paid for/reimbursed by other programs. */
  var childcare     = (client[ timeframe + 'ReasonableChildCareExpensesMonthly' ] * 12) || 0,
      childcareWork = (client[ timeframe + 'earnedIncomeBecauseOfChildcare' ] * 12) || 0;
  sum += Math.min( childcareWork, childcare );

  // ---- Adult care ---- \\
  /** @todo Only non-reimbursed
  * @todo Non-medical? */
  var nonElderlyDisabledCare = (client[ timeframe + 'NonElderlyDisabledExpenseMonthly' ] * 12) || 0,
      eldDisWork             = (client[ timeframe + 'earnedIncomeBecauseOfAdultCare' ] * 12) || 0;
  nonElderlyDisabledCare = Math.min( eldDisWork, nonElderlyDisabledCare );
  /** @todo Only non-reimbursed
  * @todo Don't include stuff from the current amount? How are we supposed to collect this data */
  var elderlyOrDisabledMedical = (client[ timeframe + 'ElderlyOrDisabledMedicalMonthly' ] * 12) || 0;

  var elderlyOrDisabledTotal      = nonElderlyDisabledCare + elderlyOrDisabledMedical,
      minElderlyOrDisabled        = elderlyOrDisabledTotal - (yearlyGross * 0.03),
      exceedingElderlyOrDisabled  = Math.max( 0, minElderlyOrDisabled );
  sum += exceedingElderlyOrDisabled;

  return sum;
};  // End getAllowancesMonthly()

/** @todo description
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
// const getNetIncomeMonthly = function ( client, timeframe ) {};  // End getNetIncomeMonthly()


// Adjusted gross?
// from 2006 ~~Source: https://www.hudexchange.info/resources/documents/incomeresidentrentcalc.pdf~~
const getAdjustedIncomeMonthly = function ( client, timeframe ) {

  // or use the function?
  var yearlyGross = client[ timeframe + 'GrossIncomeMonthly' ],
      allowances  = getAllowancesMonthly( client, timeframe ),
      disregard   = getEarnedIncomeDisregardMonthly( client, timeframe ),
      prelim      = yearlyGross - ( allowances + disregard ),
      adjusted    = Math.max( 0, prelim );

  return adjusted;
}


export {
  deductions, getSimpleGrossIncomeMonthly,
  getGrossUnearnedIncomeMonthly, getAdjustedIncomeMonthly,
  toCashflow, sumCashflow
};
