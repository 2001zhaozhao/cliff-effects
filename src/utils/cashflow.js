/** For all income and general cash flow getters and calculators
* 
* @todo Add checks for properties with messages for mistakes
* @todo Deal with expiration dates
* @todo Prune values now understood to be non-standard
*/

// DATA
import { UNEARNED_INCOME_SOURCES, CHILD_CARE_EXPENSES } from '../data/state/massachusetts/name-cores';

// ==================================
// DEDUCTIONS
// ==================================
/**
* Calculates various expenses based on form input
* Everything is monthly
*/
var expenses = {};

/** Total child and adult dependent costs (does not include child
* support paid out). Dependent = child or disabled adult
* 
* @function
* @param {object} client
* @param {string} timeframe - can be either 'current' or 'future'.
* 
* @returns {number} Total dependent care expenses
*/
expenses.dependentCare = function ( client, timeframe ) {
  var child = sumCashflow( client, timeframe, CHILD_CARE_EXPENSES ),
      /** @todo ADULT_CARE_EXPENSES */
      adult = 0; 
  return child + adult;
};  // End expenses.dependentCare()


// ==================================
// STRAIGHT UP INCOME
// ==================================

/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const getGrossUnearnedIncomeMonthly = function ( client, timeframe ) {
  return sumCashflow( client, timeframe, UNEARNED_INCOME_SOURCES );
};  // End getGrossUnearnedIncomeMonthly()


/**
* Total monthly earned and unearned income with no deductions or
* exclusions.
* 
* @function
* @param {object} client
* @param {string} timeframe - can be either 'current' or 'future'.
* 
* @returns {number} total earned and unearned monthly
* income with no deductions or exclusions.
*/
const getSimpleGrossIncomeMonthly = function ( client, timeframe ) {
  var earned    = client[ timeframe ].earned,
      unearned  = getGrossUnearnedIncomeMonthly( client, timeframe );
	return earned + unearned;
};  // End getSimpleGrossIncomeMonthly()


// ==================================
// INCOME HELPERS
// ==================================

/** @todo description
* Monthly
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const sumCashflow = function ( client, timeframe, props ) {
  var sum = 0;
  for (let namei = 0; namei < props.length; namei++) {
    sum += toCashflow( client, timeframe, props[ namei ] );  // if that prop exists, add it
  };
  return sum;
};  // End sumCashflow()


/** @todo description
* Monthly
* 
* @todo Create and normalize client object instead.
* 
* @function
* @param {object} props
* @property {object} props.__ - explanation
* 
* @returns Component
*/
const toCashflow = function ( client, timeframe, prop ) {
  return client[ timeframe ][ prop ] || 0;
};  // End toCashflow()


export {
  expenses,
  getSimpleGrossIncomeMonthly, getGrossUnearnedIncomeMonthly,
  toCashflow, sumCashflow
};
