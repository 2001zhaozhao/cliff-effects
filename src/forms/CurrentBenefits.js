// REACT COMPONENTS
import React from 'react';
import { Form } from 'semantic-ui-react';

// PROJECT COMPONENTS
import { FormPartsContainer, MassiveToggle, ControlledRadioYesNo } from './formHelpers';

// COMPONENT HELPER FUNCTIONS
import { getTimeSetter } from '../utils/getTimeSetter';


/** @todo description
*
* @todo Add "vertical list of options" creator that will create a list of fields using the `.field-aligner` class
*
* @function
* @param {object} props
* @property {object} props.__ - explanation
*
* @returns Component
*/
const CurrentBenefitsContent = ({ current, setClientProperty }) => {

  let setChecked = function ( evnt, inputProps ) {
    inputProps.checked = inputProps.label === 'yes' ? true : false;
    var obj = { ...inputProps, value: inputProps.value };
    setClientProperty( evnt, obj );
  };  // End setChecked()


  return (
    <div className={'field-aligner'}>

      <ControlledRadioYesNo
        labelText = "Do you have Section 8 Housing ?"
        checked={current.hasHousing}
        onChange={setChecked}
        name="hasHousing"
      />

      <ControlledRadioYesNo
        labelText = "Do you have Snap?"
        checked={current.hasSnap}
        onChange={setChecked}
        name="hasSnap"
      />
    </div>
  );  // end return

};  // End CurrentBenefitsContent()

/** @todo description
*
* @function
* @param {object} props
* @property {object} props.__ - explanation
*
* @returns Component
*/
// `props` is a cloned version of the original props. References broken.
const CurrentBenefitsStep = (props) => {

  const setTimeProp = getTimeSetter( 'current', props.changeClient );

  return (
    <Form size='massive' className='household-size-form flex-item flex-column'>
      <FormPartsContainer
        title     = {'Current Benefits'}
        clarifier = {'Select the benefits you currently receive.'}
        right     = {{name: 'Next', func: props.nextStep}}>
          <CurrentBenefitsContent
          setClientProperty={setTimeProp}
          current={props.client.current} />
      </FormPartsContainer>

    </Form>
  );

};  // End CurrentBenefitsStep()

export { CurrentBenefitsStep };
