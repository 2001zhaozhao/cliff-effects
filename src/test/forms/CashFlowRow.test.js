import React from 'react';
import { shallow } from 'enzyme';

import { CashFlowRow } from '../../forms/cashflow';

test('CashFlowRow should match snapshot', () => {
  const wrapper = shallow(
    <CashFlowRow
      generic='name'
      timeState={{ name: 0 }}
      onChange={ () => {} } >
      label
    </CashFlowRow>
  );
  expect(wrapper).toMatchSnapshot();
});

test('Second ManagedNumberField child should have value of timeState[ generic ]', () => {
  const monthlyVal = 200.0;
  const wrapper = shallow(
    <CashFlowRow
      generic='name'
      timeState={{ name: monthlyVal }}
      onChange={ () => {} } />
  );
  const monthlyInput = wrapper.childAt(1);
  expect(monthlyInput.prop('value')).toBe(monthlyVal);
});

test('First ManagedNumberField child should have weekly value', () => {
  const monthlyVal = 200.0;
  const wrapper = shallow(
    <CashFlowRow
      generic='name'
      timeState={{ name: monthlyVal }}
      onChange={ () => {} } />
  );
  const weeklyInput = wrapper.childAt(0);
  expect(weeklyInput.prop('value')).toBeCloseTo(monthlyVal / (4 + 1 / 3));
});

test('Third ManagedNumberField child should have yearly value', () => {
  const monthlyVal = 200.0;
  const wrapper = shallow(
    <CashFlowRow
      generic='name'
      timeState={{ name: monthlyVal }}
      onChange={ () => {} } />
  );
  const yearlyInput = wrapper.childAt(2);
  expect(yearlyInput.prop('value')).toBeCloseTo(monthlyVal * 12);
});

test('onChange gets called correctly when each value is changed', () => {
  const mockSetClientProperty = jest.fn();
  const wrapper = shallow(
    <CashFlowRow
      generic='name'
      timeState={{ name: 200.0 }}
      onChange={ mockSetClientProperty } />
  );

  const multipliers = [
    4 + 1 / 3,
    1,
    1 / 12,
  ];
  for (var i = 0; i < 3; i++) {
    const input = wrapper.childAt(i).shallow();
    const newValue = (i + 1) * 10;
    const evnt = { target: input };
    input.prop('onChange')(evnt, { value: newValue });

    expect(mockSetClientProperty.mock.calls).toHaveLength(i + 1);
    expect(mockSetClientProperty.mock.calls[ i ][ 0 ]).toBe(evnt);
    expect(mockSetClientProperty.mock.calls[ i ][ 1 ].name).toBe('name');
    expect(mockSetClientProperty.mock.calls[ i ][ 1 ].value).toBeCloseTo(newValue * multipliers[ i ]);
  }
});
