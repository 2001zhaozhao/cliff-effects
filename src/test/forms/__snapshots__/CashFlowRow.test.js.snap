// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CashFlowRow should match snapshot 1`] = `
<CashFlowContainer
  label="label"
  message={null}
  validRow={true}
>
  <ManagedNumberField
    className="cashflow-column"
    format={[Function]}
    name="name"
    onBlur={[Function]}
    otherData={
      Object {
        "interval": "weekly",
      }
    }
    store={[Function]}
    validation={[Function]}
    value={0}
  />
  <ManagedNumberField
    className="cashflow-column"
    format={[Function]}
    name="name"
    onBlur={[Function]}
    otherData={
      Object {
        "interval": "monthly",
      }
    }
    store={[Function]}
    validation={[Function]}
    value={0}
  />
  <ManagedNumberField
    className="cashflow-column"
    format={[Function]}
    name="name"
    onBlur={[Function]}
    otherData={
      Object {
        "interval": "yearly",
      }
    }
    store={[Function]}
    validation={[Function]}
    value={0}
  />
</CashFlowContainer>
`;
